[PyTorch Numeric Suite] Add weight compare API

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36186 [PyTorch Numeric Suite] Add weight compare API**

Start PyTorch Numeric Suite under PyTorch quantization and add weight compare API to it.

Differential Revision: [D20903395](https://our.internmc.facebook.com/intern/diff/D20903395/)

Commit:
[PyTorch Numeric Suite] Add weight compare API 

Start PyTorch Numeric Suite under PyTorch quantization and add weight compare API to it.

Differential Revision: [D20903395](https://our.internmc.facebook.com/intern/diff/D20903395/)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -0,0 +1,27 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+
```
Please add comments for the functions

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -0,0 +1,27 @@
+        return None
+
+
+def compare_weights(float_dict, quantized_dict):
```
This is a user facing API so comments in standard pytorch format are needed so that doc strings work correctly.  For example, look at: https://github.com/pytorch/pytorch/blob/373dc7c8efc76b604e8148d51b78191404a2f353/torch/quantization/quantize.py#L125

raghuramank100 commented on torch/quantization/_numeric_suite.py:
```diff
@@ -0,0 +1,27 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+
+def find_match(str_list, key_str, postfix):
```
Since find_match is an internal function, let us rename it to _find_match

Commit:
Update on "[PyTorch Numeric Suite] Add weight compare API "


Start PyTorch Numeric Suite under PyTorch quantization and add weight compare API to it.

Differential Revision: [D20903395](https://our.internmc.facebook.com/intern/diff/D20903395/)

[ghstack-poisoned]

hx89 commented:
Addressed comments.

Commit:
Update on "[PyTorch Numeric Suite] Add weight compare API "


Start PyTorch Numeric Suite under PyTorch quantization and add weight compare API to it.

Differential Revision: [D20903395](https://our.internmc.facebook.com/intern/diff/D20903395/)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite.py:
```diff
@@ -0,0 +1,24 @@
+        weight_dict = compare_weights(
+            annotated_conv_model.state_dict(),
+            quantized_annotated_conv_model.state_dict(),
+        )
```
Can you add a test for a simpler model, let say
``` m=nn.Conv2d(...)
 ```
Will the code work in that case? I am not sure because find_nearest_match does this:
```
        pattern2 = "".join(s2.split(".")[0:-2])
```

hx89 replied:
Seems the case you mentioned didn't show properly, could you post it again? @raghuramank100

hx89 replied:
I tried it out and it's working. If we do .join(str[0:-2]) we will get empty string.

raghuramank100 reviewed (changes requested):
Requesting one more test, with that we are ready to land

Commit:
Update on "[PyTorch Numeric Suite] Add weight compare API "


Start PyTorch Numeric Suite under PyTorch quantization and add weight compare API to it.

Differential Revision: [D20903395](https://our.internmc.facebook.com/intern/diff/D20903395/)

[ghstack-poisoned]
